### PR TITLE
feat: add use-setstate-synchronously rule

### LINKED ISSUE #1117: [New rule] use-setstate-synchronously
**Please describe what the rule should do:**

(Flutter-specific) Attempts to call `setState` after the state has been disposed usually represents a potential bug. To remedy this, the user needs to either:
1. Insert a check for `mounted` and call `setState` in it
2. Insert a guard for `!mounted` that diverges

Ideally, this rule should also be customizable to detect/forbid usages other than `setState`.

**If your rule is inspired by other please provide link to it:**

**What category of rule is this? (place an "X" next to just one item)**

- [x] Warns about a potential error (problem)
- [ ] Suggests an alternate way of doing something (suggestion)
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**

In methods:

```dart
class FooState extends State {
  fetchData() async {
    final data = await ...;
    setState(() { ... });
  }
}
```

In closures:

```dart
Widget build(context) {
  return FooWidget(
    ..
    onChange: (value) async {
      await fetchData(...);
      setState(() { ... });
    }
  );
}
```

Conditional await (would `mounted` checking be mandatory in this case?)
```dart
void fetchCacheableData() async {
  final data = someCondition ? fromCache : await ...;
  setState(() { ... });
}
```

**Are you willing to submit a pull request to implement this rule?**
Yes

### PR DESCRIPTION
**Pull request by Desdaemon**
### What is the purpose of this pull request? (put an "X" next to an item)

- [ ] Documentation update
- [ ] Bug fix
- [x] New rule
- [ ] Changes an existing rule
- [ ] Add autofixing to a rule
- [ ] Add a CLI option
- [ ] Add something to the core
- [ ] Other, please explain:

#1117

### What changes did you make? (Give an overview)

- [x] Basic implementation
- [x] Rule customization
- [x] Flow analysis
  - [x] `if` checks
  - [x] `while` checks
  - [ ] ~~Ternaries~~
  - [x] And-chains (`mounted && ..`)
  - [x] Or-chains (`!mounted || ..`)

### Is there anything you'd like reviewers to focus on?

Should this rule try to be as comprehensive as possible and check for all cases, or should the current heuristics for determining mounted-ness be sufficient?

### CODE REVIEW

**Comment by incendial** [comment]
@Desdaemon could you please update the changelog?

> Should this rule try to be as comprehensive as possible and check for all cases, or should the current heuristics for determining mounted-ness be sufficient?

The most important that the rule shouldn't produce false-positives. So, if I'm not mistaken, covering `And-chains (mounted && ..)` is a must.

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_async_setstate/avoid_async_setstate_rule.dart
Comment (incendial) [inline lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_async_setstate/avoid_async_setstate_rule.dart:16]: Thinking more about the name and looks like it can confuse people like it bans `setState(async () {});`. Maybe you have other ideas how to name it?

**Review by incendial — COMMENTED** [review]
Great result! Left a few comments. Haven't seen `@pragma('vm:prefer-inline')` in action before 🙂

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_async_setstate/avoid_async_setstate_rule.dart
Comment (Desdaemon) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_async_setstate/avoid_async_setstate_rule.dart:16]: We could follow the precedence here and call it `use-setstate-synchronously`, what do you think?

Comment (incendial) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_async_setstate/avoid_async_setstate_rule.dart:16]: Yeah, IMO, this one is better 👍

File: test/src/analyzers/lint_analyzer/rules/rules_list/use_setstate_synchronously/use_setstate_synchronously_rule_test.dart
Comment (incendial) [inline test/src/analyzers/lint_analyzer/rules/rules_list/use_setstate_synchronously/use_setstate_synchronously_rule_test.dart:17]: Please update the name here too

**Review by incendial — COMMENTED** [review]
Amazing! The only question I have - will `||` create false positives? I'll later run this rule on Flutter codebase (or you can, if you want to) to know for sure, but it'd be better if we avoid as many false positives as possible

**Comment by Desdaemon** [comment]
There's one I haven't found a way to cleanly implement:

```dart
if (!mounted || foo) return;
setState(); // should be mounted here

if (!mounted && bar || foo) return;
setState(); // LINT
```

When the first or-chain diverges, it should also have the effect of guaranteeing the widget is mounted, but currently it does not do so. If there are other false positives I'll prioritize fixing them as well.

**Comment by Desdaemon** [comment]
> I'll later run this rule on the Flutter codebase (or you can, if you want to) to know for sure, but it'd be better if we avoid as many false positives as possible

I ran this on the Flutter package on the stable branch, the only violation it detected was this situation:

https://github.com/flutter/flutter/blob/135454af32477f815a7525073027a3ff9eff1bfd/packages/flutter/lib/src/material/refresh_indicator.dart#L432-L439

I think this is too specific for this rule to handle, but otherwise no false positives yet. I'll run this on other codebases as well to cover as many edge cases as possible.

**Comment by incendial** [comment]
> I think this is too specific for this rule to handle, but otherwise no false positives yet.

Yeah, looks questionable, tbh.

> I'll run this on other codebases as well to cover as many edge cases as possible.

I've noticed a fix for a false positive, if you can also add tests - that'd be great. And after that I'm ready to merge, so tell me when you're ready.

**Comment by Desdaemon** [comment]
@incendial That should be my last batch of changes, should be ready to review and merge.

**Comment by incendial** [comment]
@Desdaemon please add tests for the config ([example](https://github.com/dart-code-checker/dart-code-metrics/blob/master/test/src/analyzers/lint_analyzer/rules/rules_list/avoid_returning_widgets/avoid_returning_widgets_rule_test.dart#L62)) and I'll merge

**Comment by incendial** [comment]
And please check the failing tests

**Review by incendial — APPROVED** [review]

**Comment by incendial** [comment]
@Desdaemon thank you for the contribution! 🚀

**Pull request merged**
